scio-test/src/main/scala/com/spotify/scio/testing/CoderAssertions.scala

```diff
@@ -30,102 +31,128 @@ import scala.reflect.ClassTag
+      coderAssertion: CoderAssertionT[T]
+    )(implicit c: Coder[T]): AssertionContext[T] = {
+      val ctx = AssertionContext(Some(value), c)
+      ctx.copy(lastAssertion = Some(coderAssertion(ctx)))
+    }
+
+    def kryoCoderShould(
+      coderAssertion: CoderAssertionT[T]
+    )(implicit ct: ClassTag[T]): AssertionContext[T] = {
+      val ctx = AssertionContext(Some(value), Coder.kryo[T])
+      ctx.copy(lastAssertion = Some(coderAssertion(ctx)))
+    }
 
-  trait CoderAssertion[T] {
-    def assert(value: T)(implicit c: Coder[T], eq: Equality[T]): Assertion
+    def coderShouldWithOpts(opts: PipelineOptions)(implicit c: Coder[T]): AssertionContext[T] =
+      AssertionContext(Some(value), c, opts = opts)
   }
 
-  def roundtripWithCustomAssert[T](
-    opts: PipelineOptions = DefaultPipelineOptions
-  )(customAssertEquality: (T, T) => Assertion): CoderAssertion[T] =
-    new CoderAssertion[T] {
-      override def assert(value: T)(implicit c: Coder[T], eq: Equality[T]): Assertion = {
-        val beamCoder = CoderMaterializer.beamWithDefault(c, o = opts)
-        val result = roundtripWithCoder(beamCoder, value)
-        customAssertEquality(value, result)
-      }
+  implicit class CoderShouldSyntax[T](c: Coder[T]) {
+    def coderShould(
+      coderAssertion: CoderAssertionT[T]
+    )(implicit c: Coder[T]): AssertionContext[T] = {
+      val ctx = AssertionContext(None, c)
+      ctx.copy(lastAssertion = Some(coderAssertion(ctx)))
     }
+  }
 
-  def roundtrip[T](
+  case class AssertionContext[T](
+    actualValue: Option[T],
+    coder: Coder[T],
+    lastAssertion: Option[Assertion] = None,
     opts: PipelineOptions = DefaultPipelineOptions
-  ): CoderAssertion[T] =
-    new CoderAssertion[T] {
-      override def assert(value: T)(implicit c: Coder[T], eq: Equality[T]): Assertion = {
-        val beamCoder = CoderMaterializer.beamWithDefault(c, o = opts)
-        checkRoundtripWithCoder(beamCoder, value)
-      }
-    }
+  ) extends AssertionContextBase {
+    override type ValType = T
 
-  def roundtripKryo[T: ClassTag](
-    opts: PipelineOptions = DefaultPipelineOptions
-  ): CoderAssertion[T] =
-    new CoderAssertion[T] {
-      override def assert(value: T)(implicit c: Coder[T], eq: Equality[T]): Assertion = {
-        val kryoCoder = CoderMaterializer.beamWithDefault(Coder.kryo[T], o = opts)
-        checkRoundtripWithCoder(kryoCoder, value)
-      }
-    }
+    def and(
+      coderAssertion: CoderAssertionT[T]
+    ): AssertionContext[T] = copy(lastAssertion = Some(coderAssertion(this)))
+  }
 
-  def notFallback[T: ClassTag](opts: PipelineOptions = DefaultPipelineOptions): CoderAssertion[T] =
-    new CoderAssertion[T] {
-      override def assert(value: T)(implicit c: Coder[T], eq: Equality[T]): Assertion = {
-        c should !==(Coder.kryo[T])
-        val beamCoder = CoderMaterializer.beamWithDefault(c, o = opts)
-        checkRoundtripWithCoder[T](beamCoder, value)
-      }
-    }
+  trait AssertionContextBase {
```

Comment:
need this trait without type parameter to support case like:
```
def beOfType[ExpectedCoder: ClassTag](): CoderAssertion
```
to avoid passing two type parameters